packages/yarnpkg-core/sources/Configuration.ts

```diff
@@ -1126,23 +1143,40 @@ export class Configuration {
         if (!Array.isArray(data.plugins))
           continue;
 
+        const missingPluginMetaList = [];
         for (const userPluginEntry of data.plugins) {
           const userProvidedPath = typeof userPluginEntry !== `string`
             ? userPluginEntry.path
             : userPluginEntry;
+          const userProvidedSpec = userPluginEntry?.spec ?? ``;
 
           const pluginPath = ppath.resolve(cwd, npath.toPortablePath(userProvidedPath));
+          if (!await xfs.existsPromise(pluginPath)) {
+            if (!userProvidedSpec.match(/^https?:/)) {
+              const prettyPluginName = formatUtils.pretty(configuration, ppath.basename(pluginPath, `.cjs`), formatUtils.Type.NAME);
+              const prettyGitIgnore = formatUtils.pretty(configuration, `.gitignore`, formatUtils.Type.NAME) ;
+              const prettyYarnrc = formatUtils.pretty(configuration, configuration.values.get(`rcFilename`), formatUtils.Type.NAME) ;
+              const prettyUrl = formatUtils.pretty(configuration, `https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored`, formatUtils.Type.URL) ;
+              throw new UsageError(`Failed to get ${prettyPluginName} plugin, please try to delete the plugin from ${prettyYarnrc} and reinstall it, the error usually occurs because ${prettyGitIgnore} is incorrect, please check ${prettyUrl} to set the correct ${prettyGitIgnore}`);
+            }
+
+            const pluginBuffer = await httpUtils.get(userProvidedSpec, {configuration});
```

Comment:
We should check here that this buffer's checksum matches the expectation (and add a test)

Reply:
I was also thinking where we should add an option like `--check cache`, but I can't think of one.

Should we just throw an error by default if the checksum is changed?

Reply:
Yes, I think so - if users don't want them to be checked for consistency they can either 1/ remove the `checksum` from the rc files (perhaps we can make that an option in `plugin import --no-checksum`) 2/ check-in the plugins in their repos

Reply:
> We should check here that this buffer's checksum matches the expectation (and add a test)

OK~~

https://github.com/yarnpkg/berry/pull/4912/commits/41c3a6e47a8a410978d437e403876939bb4e599a
https://github.com/yarnpkg/berry/pull/4912/commits/f2a3f9997285aad197fc7d75f5a5d7a2cb171009

---

I have added a new option and a new command

#### `plugin import --no-checksum`

If there are no checksum field in the RC file, it will not check for changes when Configuration fetch the missing plugin.

There are two possibilities without checksum field:
- plugin installed before this PR
- plugin installed with this new command

When the plugin already has checksum, running this command will update the plugin and clear the plugin's checksum

https://github.com/yarnpkg/berry/pull/4912/commits/13a86972b106d19691ebc6aa8127dae98ed63563
https://github.com/yarnpkg/berry/pull/4912/commits/03b19cf5d4d5cf79bd86de2ee02fa6bd94ca4c6b

#### `plugin check`

Compare local checksums with remote checksums (all installed plugins), then print all changed plugins, when used in a CI environment, it throws an error
 
I think this will allow users to avoid the risk of accepting third party PR update plugins
 
https://github.com/yarnpkg/berry/pull/4912/commits/f53e87f45f4627f79d79eafea0902b6431329728
https://github.com/yarnpkg/berry/pull/4912/commits/2218016adf02d46641353688a54f415e75ff9f57